vispy/visuals/volume.py

```diff
@@ -114,6 +116,11 @@
     return color1.g; // todo: why did I have this abstraction in visvis?
 }}
 
+vec4 applyColormap(float data) {{
+    float val = max(data - clim.x, 0) / (clim.y - clim.x);
```

Comment:
Are users allowed to have flipped color limits? It is common in the data I work with to flip the color limits so that `clim = (50, 25)` and the data is then inverted. Does this `max` interfere with that? Is it needed? Is it this way in the ImageVisual PR that you made?

Reply:
hmm, good question.  Does the existing vispy normalization code work for your application?  Yes, I believe there is a `max()` call in the ImageVisual PR as well.  But I can submit a followup PR if we identify the desired behavior.

For what it's worth, if I run the `examples/basics/scene/volume.py` in the master branch, with a flipped `clim` argument (.e.g `(1000, 0)`), it doesn't work: the volume is all white.  Same goes for the image example, even if I revert back before the ImageVisual PR, like a22df65b.  Is this something you often do with vispy?  or a new feature that would be nice to support?

Reply:
It's possible I have it custom coded in my own application since I'm using floats there:

https://github.com/ssec/sift/blob/4874d971811b16502b23b3837069ed25cd454942/uwsift/view/visuals.py#L147-L153

To support it you'd have to make sure the python code checks that the new clims are within the entire previous clim range instead of just `new_clim_max > old_clim_max`.

As for this PR, is the `max` doing anything?

Reply:
yeah, on this PR the max is making sure that we don't get wraparound... Without it, if you bring the clim min above the minimum values in the image, those pixels don't go black, they get bright again

Reply:
Thoughts on switching to what I'm doing at the above linked code where I use `clamp` to clip the data to the provided color limits?

Reply:
looks good to me! will work on fixing this PR to match it, probably tomorrow.

Reply:
hey @djhoese, sorry for the long delay, though I gather you've been busy too :)

I updated the code to match your clamp pattern instead of using max.  If it's ok with you, I'd propose that the clims inverting behavior be addressed in a separate PR, since that's a new feature altogether.  Let me know if you disagree, or if you see anything else you'd like addressed in this PR.

Reply:
> hey @djhoese, sorry for the long delay, though I gather you've been busy too :)
> 

@tlambert03 FYI, David is AFK for two weeks or so. I can have a look into this PR by mid of next week. We might have to wait until David is back.

Reply:
@tlambert03 Thanks for the updates. Correct me if I'm wrong, but don't your recent changes already handle the clim inversion?

Reply:
I don't think so?  at least, not in the sense that providing clims (1, 0) will give you an inverted lookup-table.  I thought that was also going to require some changes to `Volume.set_data`, which I mostly hadn't touched in this PR:

https://github.com/vispy/vispy/blob/77f016a0c6121e1b0c024bdf4e03d5bf84aa9879/vispy/visuals/volume.py#L486-L493

it wouldn't be too hard to support... but that's why I was suggesting that we maybe address that in a follow-up PR.  but let me know if you want it all done here and now

Reply:
Ah I think you're right. I was thinking of my own application where I have floating point data in the texture and there is not CPU handling of the clims. I'd like it in this PR, but if it is too complex I am fine moving it to another PR. I'd really like it to be included in the next release (I don't want the existing functionality released without the inverse clims possible).

Reply:
works for me.  let's leave this PR open then as motivation :) I'll work on it soon

Reply:
ok just pushed an update with [inversion](https://github.com/vispy/vispy/pull/1842/commits/1c71091c0be60fd03c34a3dee7cba18657302ce3).  I have to say, my brain got a little turned around... and while I think this code achieves the correct behavior (in my tests), I may have made it more complicated than it needs to be, and would appreciate a critical review.  

Also, adding support for inversion  _and_ shader-based clim changes definitely seems to increase the possibility of bugs if someone switches from a standard clim to an inverted one after instantiating the volume...  not 100% I've caught all of those.